workflow/util/util.go

```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
+type deleteFn func(string, bool)
+type tillFn func(*node) (bool, bool)
+
+func getTillFnNodeType(nodeType wfv1.NodeType) tillFn {
+	return func(n *node) (bool, bool) {
+		return n.n.Type == nodeType, true
+	}
 }
 
-func isGroupNode(node wfv1.NodeStatus) bool {
-	return node.Type == wfv1.NodeTypeDAG || node.Type == wfv1.NodeTypeTaskGroup || node.Type == wfv1.NodeTypeStepGroup || node.Type == wfv1.NodeTypeSteps
+func consumeTill(n *node, should tillFn, resetFunc resetFn) (*node, error) {
+	curr := n
+	for {
+		if curr == nil {
+			return nil, fmt.Errorf("was seeking node but ran out of nodes to explore")
+		}
+
+		if foundNode, shouldReset := should(curr); foundNode {
+			if shouldReset {
+				resetFunc(curr.n.ID, true)
+			}
+			return curr, nil
+		}
+		curr = curr.parent
+	}
+}
+
+func getTillBoundaryFn(boundaryID string) tillFn {
+	return func(n *node) (bool, bool) {
+		return n.n.ID == boundaryID, n.n.BoundaryID != ""
+	}
 }
 
-func resetConnectedParentGroupNodes(oldWF *wfv1.Workflow, newWF *wfv1.Workflow, currentNode wfv1.NodeStatus, resetParentGroupNodes []string) (*wfv1.Workflow, []string) {
-	currentNodeID := currentNode.ID
+func consumeBoundaries(n *node, resetFunc resetFn) (*node, error) {
+	curr := n
 	for {
-		currentNode, err := oldWF.Status.Nodes.Get(currentNodeID)
+		if curr == nil {
+			return curr, nil
+		}
+		if curr.parent != nil && curr.parent.n.Type == wfv1.NodeTypeStepGroup {
+			resetFunc(curr.parent.n.ID, true)
+		}
+		seekingBoundaryID := curr.n.BoundaryID
+		if seekingBoundaryID == "" {
+			return curr.parent, nil
+		}
+		var err error
+		curr, err = consumeTill(curr, getTillBoundaryFn(seekingBoundaryID), resetFunc)
 		if err != nil {
-			log.Panicf("dying due to inability to obtain node for %s, panicking", currentNodeID)
+			return nil, err
+		}
+	}
+}
+
+func consumeStepGroup(n *node, resetFunc resetFn) (*node, error) {
+	return consumeTill(n, getTillFnNodeType(wfv1.NodeTypeStepGroup), resetFunc)
+}
+
+func consumeSteps(n *node, resetFunc resetFn) (*node, error) {
+	n, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypeSteps), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	return consumeBoundaries(n, resetFunc)
+}
+
+func consumeTaskGroup(n *node, resetFunc resetFn) (*node, error) {
+	return consumeTill(n, getTillFnNodeType(wfv1.NodeTypeTaskGroup), resetFunc)
+}
+
+func consumeDAG(n *node, resetFunc resetFn) (*node, error) {
+	n, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypeDAG), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	return consumeBoundaries(n, resetFunc)
+}
+
+func consumePod(n *node, resetFunc resetFn, addToDelete deleteFn) (*node, error) {
+	// this sets to reset but resets are overridden by deletes in the final FormulateRetryWorkflow logic.
+	curr, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypePod), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	addToDelete(curr.n.ID, true)
+	children := getChildren(curr)
+	for childID := range children {
+		addToDelete(childID, true)
```

Comment:
This recurses one child only, which is surprising... one child might be correct today but it might be a fun source of orphans in the future. This is not a sentence I expected to write.

Ah, I suspect that new line 1346 prevents us keeping children of deleted nodes anyway, so this whole child deletion loop could be removed instead.

Reply:
> This recurses one child only 

I don't think so ? We get all children of curr. The getChildren does a BFS to obtain all the children 

> Ah, I suspect that new line 1346 prevents us keeping children of deleted nodes anyway, so this whole child deletion loop could be removed instead.

Yes but we shouldn't rely on that kind of inference in my opinion. For all other executable node types, we need to track the deletion id so that we can delete the pods. We shouldn't diverge from this logic just because it happens not  to be needed for containersets.

Reply:
Ok, I follow and agree now. Thanks.